rfcs/0158-artifact-metadata.md

```diff
@@ -0,0 +1,62 @@
+# RFC 158 - artifact metadata
+* Comments: [#158](https://github.com/taskcluster/taskcluster-rfcs/pull/158)
+* Proposed by: @escapewindow
+
+# Summary
+
+The goal is to provide Artifact Integrity guarantees from the point the worker uploads an artifact, to the point that someone downloads that artifact to use. We can do this by:
+
+1. adding SHA metadata to artifacts in the Queue,
+2. ensuring that that metadata can't be modified once it's written, and
+3. providing a download tool that queries the Queue for an artifact's location and SHA, downloads the artifact, and verifies the downloaded SHA matches the SHA provided by the Queue.
+
+## Motivation
+
+This would improve robustness and security. By storing a SHA to verify on download, we can avoid corrupt downloads. By verifying the read-only SHA before we use the artifact, we can detect malicious tampering of artifacts at rest before we use it for a critical operation.
+
+(This would obsolete the [artifacts section of the Chain of Trust artifact](https://scriptworker.readthedocs.io/en/latest/chain_of_trust.html#chain-of-trust-artifacts), and improve artifact integrity platform-wide.)
+
+# Details
+
+## Adding artifact metadata to the Queue
+
+First, we add a `metadata` dictionary to the `S3ArtifactRequest` type. This is a dictionary to allow for flexibility of usage. The initial known keys would include
```

Comment:
Does Taskcluster handle or proxy in some way this request to S3? Can Taskcluster itself add these metadata fields rather than having them be user provided?

Reply:
Aiui, Taskcluster never sees the artifact.

https://github.com/taskcluster/taskcluster/blob/4441d319a79a657b392d6c903686f166f3636da1/ui/docs/reference/platform/queue/worker-interaction.mdx#uploading-artifacts

Upload:

- worker calls `queue.createArtifact` to get an s3 `putUrl` (currently without sha metadata; RFC suggests adding sha metadata)
- worker uploads to s3 via the `putUrl`

Download:
- worker calls [`queue.getArtifact`](https://github.com/taskcluster/taskcluster/blob/99906758c617ae0d9205bd0c3e20b220d227262b/clients/client-py/taskcluster/generated/queue.py#L394-L477) or [`buildSignedUrl`](https://github.com/taskcluster/taskcluster/blob/99906758c617ae0d9205bd0c3e20b220d227262b/clients/client-py/taskcluster/client.py#L179-L244), which redirects or points directly at S3.
- (RFC suggests we also query for artifact metadata and verify, via a download tool)

Reply:
However, we've noted that if we don't trust the worker to generate trusted SHAs, we also can't trust the worker to generate trusted artifacts. So the worker is trusted; the shas mainly prevent modifying artifacts at rest.

Reply:
Ok, sounds good.